docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,79 @@
+# Qualifying replicas for read operations
+
+## Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+a replica which is overloaded or dead.
+
+## Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple estimation
+of how responsive a replica is likely to be is implemented as follows:
+ * each time a read request is sent to a replica, a timestamp is recorded
+ * each time a response is received from a replica, a timestamp is recorded
+ * each time a response is received from a replica, the request round trip time
+   is recorded
+
+The difference between the two above timestamps is a cheap, rough implementation
+of a failure detector - if the node hasn't responded for more than X milliseconds since
+the previous request and the current request is going to time out in X milliseconds,
+it gets more and more likely to be dead.
+The last recorded round trip time is, in turn, a cheap, rough estimate of how long
+it may take the replica to respond to the next request - assuming that the conditions
```

Comment:
Note to self: also bad, because a momentary spike in latency can disqualify a replica for a couple of seconds. Something more resistant to spikes is needed (moving average?)

Reply:
In https://github.com/scylladb/scylla/issues/8964 I tried to explain how what this mechanism is supposed to measure (the duration in which we got **no response**) is not actually related to latency, but rather to throughput - even when latencies are getting high, you don't normally have a very long duration of getting no responses at all.
Please tell me if I'm misunderstanding something....